Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.cpp

```diff
@@ -161,30 +119,105 @@ MacroscopicProperties::InitData ()
+    m_sigma_mf = std::make_unique<MultiFab>(ba, dmap, 1, ng_EBSolver);
+    // epsilon is cell-centered MultiFab
+    m_eps_mf = std::make_unique<MultiFab>(ba, dmap, 1, ng_EBSolver);
+    // mu is cell-centered MultiFab
+    m_mu_mf = std::make_unique<MultiFab>(ba, dmap, 1, ng_EBSolver);
+    // Initialize sigma
+    if (m_sigma_s == "constant") {
+
+        m_sigma_mf->setVal(m_sigma);
+
+    } else if (m_sigma_s == "parse_sigma_function") {
 
+        InitializeMacroMultiFabUsingParser(m_sigma_mf.get(), m_sigma_parser->compile<3>(), lev);
+    }
+    // Initialize epsilon
+    if (m_epsilon_s == "constant") {
+
+        m_eps_mf->setVal(m_epsilon);
+
+    } else if (m_epsilon_s == "parse_epsilon_function") {
+
+        InitializeMacroMultiFabUsingParser(m_eps_mf.get(), m_epsilon_parser->compile<3>(), lev);
+
+    }
+    // Initialize mu
+    if (m_mu_s == "constant") {
+
+        m_mu_mf->setVal(m_mu);
+
+    } else if (m_mu_s == "parse_mu_function") {
+
+        InitializeMacroMultiFabUsingParser(m_mu_mf.get(), m_mu_parser->compile<3>(), lev);
+
+    }
+
+    amrex::IntVect sigma_stag = m_sigma_mf->ixType().toIntVect();
+    amrex::IntVect epsilon_stag = m_eps_mf->ixType().toIntVect();
+    amrex::IntVect mu_stag = m_mu_mf->ixType().toIntVect();
+    amrex::IntVect Ex_stag = warpx.getEfield_fp(0,0).ixType().toIntVect();
+    amrex::IntVect Ey_stag = warpx.getEfield_fp(0,1).ixType().toIntVect();
+    amrex::IntVect Ez_stag = warpx.getEfield_fp(0,2).ixType().toIntVect();
 
     for ( int idim = 0; idim < AMREX_SPACEDIM; ++idim) {
+        sigma_IndexType[idim]   = sigma_stag[idim];
+        epsilon_IndexType[idim] = epsilon_stag[idim];
+        mu_IndexType[idim]      = mu_stag[idim];
         Ex_IndexType[idim]      = Ex_stag[idim];
         Ey_IndexType[idim]      = Ey_stag[idim];
         Ez_IndexType[idim]      = Ez_stag[idim];
-        Bx_IndexType[idim]      = Bx_stag[idim];
-        By_IndexType[idim]      = By_stag[idim];
-        Bz_IndexType[idim]      = Bz_stag[idim];
+        macro_cr_ratio[idim]    = 1;
     }
 #if (AMREX_SPACEDIM==2)
+        sigma_IndexType[2]   = 0;
+        epsilon_IndexType[2] = 0;
+        mu_IndexType[2]      = 0;
         Ex_IndexType[2]      = 0;
         Ey_IndexType[2]      = 0;
         Ez_IndexType[2]      = 0;
-        Bx_IndexType[2]      = 0;
-        By_IndexType[2]      = 0;
-        Bz_IndexType[2]      = 0;
+        macro_cr_ratio[2]    = 1;
 #endif
+}
+
+void
+MacroscopicProperties::InitializeMacroMultiFabUsingParser (
+                       amrex::MultiFab *macro_mf,
+                       amrex::ParserExecutor<3> const& macro_parser,
+                       const int lev)
+{
+    auto& warpx = WarpX::GetInstance();
+    const auto dx_lev = warpx.Geom(lev).CellSizeArray();
+    const RealBox& real_box = warpx.Geom(lev).ProbDomain();
+    IntVect iv = macro_mf->ixType().toIntVect();
+    for ( MFIter mfi(*macro_mf, TilingIfNotGPU()); mfi.isValid(); ++mfi ) {
```

Comment:
Hope the replacements for `auto` work:
```suggestion
    WarpX& warpx = WarpX::GetInstance();
    const amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> dx_lev = warpx.Geom(lev).CellSizeArray();
    const amrex::RealBox& real_box = warpx.Geom(lev).ProbDomain();
    amrex::IntVect iv = macro_mf->ixType().toIntVect();
    for ( amrex::MFIter mfi(*macro_mf, TilingIfNotGPU()); mfi.isValid(); ++mfi ) {
```